src/test/scala/com/cognite/spark/datasource/TimeSeriesUpsertTest.scala

```diff
@@ -0,0 +1,114 @@
+    sourceDf.cache()
+
+    val destinationDf = spark.read.format("com.cognite.spark.datasource")
+      .option("apiKey", writeApiKey)
+      .option("type", "timeseries")
+      .load()
+    destinationDf.createOrReplaceTempView("destinationTimeSeries")
+
+    // Insert time series data
+    spark.sql(
+      s"""
+         |select $initialDescription as description,
+         |name,
+         |isString,
+         |metadata,
+         |'test' as unit,
+         |'' as assetId,
+         |isStep,
+         |securityCategories,
+         |id,
+         |createdTime,
+         |lastUpdatedTime
+         |from sourceTimeSeries
+         |limit 5
+     """.stripMargin)
+      .select(destinationDf.columns.map(col): _*)
+      .write
+      .insertInto("destinationTimeSeries")
+
+
+    // Check if post worked
+    val initialDescriptionsAfterPost = retryWhile[DataFrame](
+      spark.sql(s"""select * from destinationTimeSeries where description = $initialDescription"""),
+      df => df.count < 5)
+    assert(initialDescriptionsAfterPost.count == 5)
+
+    // Upsert time series data
+    spark.sql(
+      s"""
+         |select $updatedDescription as description,
+         |name,
+         |isString,
+         |metadata,
+         |'test' as unit,
+         |'' as assetId,
+         |isStep,
+         |securityCategories,
+         |id,
+         |createdTime,
+         |lastUpdatedTime
+         |from sourceTimeSeries
+         |limit 5
+     """.stripMargin)
+      .select(destinationDf.columns.map(col): _*)
+      .write
+      .insertInto("destinationTimeSeries")
+
+    // Check if upsert worked
+    val updatedDescriptionsAfterUpsert = retryWhile[DataFrame](
+      spark.sql(s"""select * from destinationTimeSeries where description = $updatedDescription"""),
+      df => df.count < 5)
+    assert(updatedDescriptionsAfterUpsert.count == 5)
+
+    val initialDescriptionsAfterUpsert = retryWhile[DataFrame](
+      spark.sql(s"""select * from destinationTimeSeries where description = $initialDescription"""),
+      df => df.count > 0)
+    assert(initialDescriptionsAfterUpsert.count == 0)
+
+    val namesDf = retryWhile[DataFrame](
+      spark.sql(s"""select name from destinationTimeSeries where unit = 'test'"""),
+      df => df.isEmpty)
+    val namesList = namesDf.rdd.map(r => r.getAs[String](0)).collect().toList
+    cleanupTimeSeries(namesList)
+
+    val testTimeSeriesAfterCleanup = retryWhile[DataFrame](
+      spark.sql(s"""select * from destinationTimeSeries where unit = 'test'"""),
+      df => df.count > 0)
+    assert(testTimeSeriesAfterCleanup.count == 0)
+  }
+
+  def cleanupTimeSeries(names: List[String]): Unit = {
+    val project = getProject(writeApiKey, Constants.DefaultMaxRetries, Constants.DefaultBaseUrl)
+
+    for (name <- names) {
+      delete(
+        writeApiKey,
+        uri"${Constants.DefaultBaseUrl}/api/0.5/projects/$project/timeseries/$name",
+        maxRetries = 10
```

Comment:
Can use a lower value here, I think. Otherwise it takes a long time for the tests to fail when we do some thing wrong, and if we've failed to delete a time series five times it probably won't help to try another five times.